pkg/verification/service/verification_service.go

```diff
@@ -0,0 +1,265 @@
+
+		// Generate the verification message with the new verification code
+		content := fmt.Sprintf(s.config.GetVerificationMessageTemplate(), verificationCode)
+		client := twilio.NewClient(s.config.GetTwilioAccountSID(), s.config.GetTwilioAuthToken(), s.HttpClient)
+		msg, err := client.Messages.SendMessage(s.config.GetTwilioFromNumber(), e164PhoneNumber, content, nil)
+		if err != nil {
+			log.Error(ctx, err, fmt.Sprintf("error while sending, code: %d message: %s", msg.ErrorCode, msg.ErrorMessage))
+
+			// If we get an error here then just die, don't bother updating the UserSignup
+			return errors.NewInternalError(err, "error while sending verification code")
+		}
+	}
+
+	_, updateErr := s.Services().SignupService().UpdateUserSignup(signup)
+	if updateErr != nil {
+		log.Error(ctx, err, "error while updating UserSignup resource")
+		initError = errors.NewInternalError(updateErr, "error while updating UserSignup resource")
+	}
+
+	return
+}
+
+func generateVerificationCode() string {
+	buf := make([]byte, codeLength)
+	rand.Read(buf)
+
+	charsetLen := len(codeCharset)
+	for i := 0; i < codeLength; i++ {
+		buf[i] = codeCharset[int(buf[i])%charsetLen]
+	}
+
+	return string(buf)
+}
+
+// VerifyCode validates the user's phone verification code.  It updates the specified UserSignup value, so even
+// if an error is returned by this function the caller should still process changes to it
+func (s *ServiceImpl) VerifyCode(ctx *gin.Context, userID string, code string) (verificationErr error) {
+
+	// If we can't even find the UserSignup, then die here
+	signup, lookupErr := s.Services().SignupService().GetUserSignup(userID)
+	if lookupErr != nil {
+		if apierrors.IsNotFound(lookupErr) {
+			log.Error(ctx, lookupErr, "usersignup not found")
+			return errors.NewNotFoundError(lookupErr, "user not found")
+		}
+		log.Error(ctx, lookupErr, "error retrieving usersignup")
+		return errors.NewInternalError(lookupErr, fmt.Sprintf("error retrieving usersignup: %s", userID))
+	}
+
+	now := time.Now()
+
+	// If 24 hours has passed since the verification timestamp, then reset the timestamp and verification attempts
+	ts, parseErr := time.Parse(TimestampLayout, signup.Annotations[v1alpha1.UserSignupVerificationTimestampAnnotationKey])
+	if parseErr != nil || (parseErr == nil && now.After(ts.Add(24*time.Hour))) {
+		// Set a new timestamp
+		signup.Annotations[v1alpha1.UserSignupVerificationTimestampAnnotationKey] = now.Format(TimestampLayout)
+		signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = "0"
+	}
+
+	attemptsMade, convErr := strconv.Atoi(signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey])
+	if convErr != nil {
+		// We shouldn't get an error here, but if we do, we will set verification attempts to max allowed
+		// so that we at least now have a valid value, and let the workflow continue to the
+		// subsequent attempts check
```

Comment:
Same here. Maybe it's worth logging the error just in case.

Reply:
Log message added.